packages/fxa-settings/src/pages/Signin/SigninTotpCode/container.tsx

```diff
@@ -0,0 +1,156 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { RouteComponentProps, useNavigate, useLocation } from '@reach/router';
+import { useValidatedQueryParams } from '../../../lib/hooks/useValidate';
+import { SigninQueryParams } from '../../../models/pages/signin';
+import { SigninTotpCode } from './index';
+import { gql, useMutation } from '@apollo/client';
+import GleanMetrics from '../../../lib/glean';
+import { logViewEvent } from '../../../lib/metrics';
+import { hardNavigate } from 'fxa-react/lib/utils';
+import {
+  AuthUiError,
+  AuthUiErrorNos,
+  AuthUiErrors,
+} from '../../../lib/auth-errors/auth-errors';
+import { currentAccount, sessionToken } from '../../../lib/cache';
+import { MozServices } from '../../../lib/types';
+import VerificationMethods from '../../../constants/verification-methods';
+import VerificationReasons from '../../../constants/verification-reasons';
+import { GraphQLError } from 'graphql';
+
+export const viewName = 'signin-totp-code';
+
+export type SigninTotpCodeContainerProps = {
+  serviceName: MozServices;
+};
+
+export const SigninTotpCodeContainer = ({
+  serviceName,
+}: SigninTotpCodeContainerProps & RouteComponentProps) => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const storedLocalAccount = currentAccount();
+
+  // TBD: A bit of a work around. In the previous signup step
+  // a new session token should have been generated. Setting
+  // it here to ensure that the authorization header is set
+  // when it comes time to verify the totp code.
+  if (storedLocalAccount?.sessionToken) {
+    sessionToken(storedLocalAccount?.sessionToken);
+  }
+
+  const { queryParamModel } = useValidatedQueryParams(SigninQueryParams);
+  const [verifyTotpCode] = useMutation(gql`
+    mutation VerifyTotp($input: VerifyTotpInput!) {
+      verifyTotp(input: $input) {
+        success
+      }
+    }
+  `);
+
+  const handleNavigation = () => {
+    // Note: `redirectTo` looks similar to `redirectPathname`, but they
+    // work slightly differently: `redirectTo` redirects automatically,
+    // while `redirectPathname` redirects after a form submission.
+
+    if (queryParamModel.redirectToUrl) {
+      hardNavigate(queryParamModel.redirectToUrl);
+    } else if (queryParamModel.redirectPathname) {
+      navigate(queryParamModel.redirectPathname + location.search);
+    } else if (
+      storedLocalAccount?.verificationReason ===
+      VerificationReasons.CHANGE_PASSWORD
+    ) {
+      navigate(`/post_verify/password/force_password_change${location.search}`);
+    } else {
+      navigate('/settings');
+    }
+  };
+
+  const submitTotpCode = async (code: string) => {
+    try {
+      const result = await verifyTotpCode({
+        variables: {
+          input: {
+            code,
+            service: queryParamModel.service,
+          },
+        },
+      });
+      GleanMetrics.loginConfirmation.submit();
```

Comment:
FWIW I've been leaving these Glean events in page components. If you think they feel better in the container component that's OK with me, we might want an issue filed to move them over. I _think_ we'll still have some glean events in the page component either way though.